### PR TITLE
Refactor cli

### LINKED ISSUE #895: Refactor: mythril/interfaces/cli
## Description
This issue tracks maintenance of the mythril/interfaces/cli.py

One change that we might want to introduce is the use of commands, for example
`mythril analyze contract.sol`

As these provide intuitive ways to separate the use of different flags

## Checkpoints:
- [ ] Cleanup Code
- [ ] Ensure 80% code coverage
- [ ] Ensure all public functions have been outfitted with the proper documentation

### PR DESCRIPTION
**Pull request by norhh**
fixes #895.

### CODE REVIEW

**Comment by norhh** [comment]
I don't feel it easy to extend it based on the feedback in the docs, I should re-refactor this.

File: mythril/interfaces/cli.py [hunk 1]
```diff
@@ -53,138 +58,251 @@ def exit_with_error(format_, message):
...
         "-o",
         "--outform",
         choices=["text", "markdown", "json", "jsonv2"],
         default="text",
         help="report output format",
         metavar="<text/markdown/json/jsonv2>",
     )
-    outputs.add_argument(
+    parser.add_argument(
         "--verbose-report",
```
Comment (nbanmp) [inline mythril/interfaces/cli.py]: The `-o/--outform` and `--verbose-report` arguments are displayed for disassemble, read-storage, leveldb-search, function-to-hash,  hash-to-address, and even version. They are only needed for `analyze`.

File: mythril/interfaces/cli.py [hunk 2]
```diff
@@ -475,26 +605,21 @@ def execute_command(
                     args.outform, "Error loading analyis modules: " + format(e)
```
Comment (nbanmp) [inline mythril/interfaces/cli.py]: This should be `analysis`

File: mythril/interfaces/cli.py [hunk 3]
```diff
@@ -507,35 +632,40 @@ def parse_args(parser: argparse.ArgumentParser, args: argparse.Namespace) -> Non
         os.system(" ".join(sys.argv) + " | python3 " + path + "/epic.py")
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/interfaces/cli.py]: Errors aren't stdout right? we can't put them in epic.

[the remark this answers is not in the document]
Comment (norhh) [reply mythril/interfaces/cli.py]: Also, epic usually seems to work. except when only using epic :laughing:

File: mythril/interfaces/cli.py (hunk 1 above)
Comment (norhh) [reply mythril/interfaces/cli.py]: `and even version` . json output stuff is being used in version.(but not the rest)

File: mythril/interfaces/cli.py [hunk 4]
```diff
@@ -53,138 +58,251 @@ def exit_with_error(format_, message):
...
     )
 
-    utilities = parser.add_argument_group("utilities")
-    utilities.add_argument(
-        "--hash", help="calculate function signature hash", metavar="SIGNATURE"
+    subparsers = parser.add_subparsers(dest="command", help="Commands")
+    for analyze_string in ANALYZE_LIST:
+        analyzer_parser = subparsers.add_parser(
+            analyze_string,
+            help="Triggers the analysis of the smart contract",
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/interfaces/cli.py]: That's cool, I searched for something like this, but couldn't find this one. Now code won't be as terrible

**Comment by norhh** [comment]
`In addition, it would be good to print the help for the command used when an invalid argument is passed in:` It can be done, but not to a that good extent. The parser isn't sure whether the flag belongs to parser or subparser, so it outputs all in those cases.

File: mythril/interfaces/cli.py [hunk 5]
```diff
@@ -387,54 +511,54 @@ def get_code(disassembler: MythrilDisassembler, args: argparse.Namespace):
         )  # list of files
     else:
         exit_with_error(
```
Comment (nbanmp) [inline mythril/interfaces/cli.py]: This error is reached when myth is ran without any args.
```
━━┫ ./myth                       
usage: myth [-h] [-v LOG_LEVEL]
            {analyze,a,disassemble,d,read-storage,leveldb-search,function-to-hash,hash-to-address,version,truffle,help}
            ...
mythril.interfaces.cli [ERROR]: No input bytecode. Please provide EVM code via -c BYTECODE, -a ADDRESS, -f BYTECODE_FILE or <SOLIDITY_FILE> 
```

File: mythril/interfaces/cli.py [hunk 6]
```diff
@@ -472,77 +611,83 @@ def execute_command(
...
+    sys.exit()
+
+
+def parse_args_and_execute(parser: ArgumentParser, args: Namespace) -> None:
     """
     Parses the arguments
     :param parser: The parser
     :param args: The args
     """
+    if args.command not in COMMAND_LIST:
```
Comment (nbanmp) [inline mythril/interfaces/cli.py]: This should be below `if args.epic`

```
━━┫ ./myth --epic                
usage: myth [-h] [-v LOG_LEVEL]
            {analyze,a,disassemble,d,read-storage,leveldb-search,function-to-hash,hash-to-address,version,truffle,help}
            ...
mythril.interfaces.cli [ERROR]: No input bytecode. Please provide EVM code via -c BYTECODE, -a ADDRESS, -f BYTECODE_FILE or <SOLIDITY_FILE> 
<Everything after this is "epic">
usage: myth [-h] [-v LOG_LEVEL]
            {analyze,a,disassemble,d,read-storage,leveldb-search,function-to-hash,hash-to-address,version,truffle,help}
            ...
```

File: mythril/interfaces/cli.py [hunk 7]
```diff
@@ -472,77 +611,83 @@ def execute_command(
...
             sys.exit()
 
-        address = get_code(disassembler, args)
+        address = load_code(disassembler, args)
         execute_command(
             disassembler=disassembler, address=address, parser=parser, args=args
         )
     except CriticalError as ce:
-        exit_with_error(args.outform, str(ce))
+        exit_with_error(args.__dict__.get("outform", "text"), str(ce))
```
Comment (nbanmp) [inline mythril/interfaces/cli.py]: Everything else just passes in `args.outform` directly. Is there a reason these `exit_with_error`s are different from the other ones?

File: mythril/interfaces/cli.py [hunk 8]
```diff
@@ -387,54 +511,54 @@ def get_code(disassembler: MythrilDisassembler, args: argparse.Namespace):
...
-        create_timeout=args.create_timeout,
-        enable_iprof=args.enable_iprof,
-        disable_dependency_pruning=args.disable_dependency_pruning,
-        onchain_storage_access=not args.no_onchain_storage_access,
-    )
-
-    if args.disassemble:
-        # or mythril.disassemble(mythril.contracts[0])
-
+    if args.command in DISASSEMBLE_LIST:
```
Comment (nbanmp) [inline mythril/interfaces/cli.py]: this can be elif, then we don't wont need the return above this either.

File: mythril/interfaces/cli.py (hunk 5 above)
Comment (norhh) [reply mythril/interfaces/cli.py]: Oh! This should be added as a test.

File: mythril/interfaces/cli.py (hunk 7 above)
Comment (norhh) [reply mythril/interfaces/cli.py]: args.outform may not exist here.

**Comment by nbanmp** [comment]
We also will need to update the docs with the new syntax - https://mythril-classic.readthedocs.io/en/master/security-analysis.html Do we want to do that in this PR or another one?

**Comment by norhh** [comment]
Next one

**Pull request merged**